[InstCombine] Add support for Instruction combining of hyperbolic functions

kshitijvp opened the pull request:
Fixes llvm/llvm-project#173706

kshitijvp committed:
Precommit Tests

kshitijvp committed:
Added support for hyperbolic functions

Added pattern matching for all hyperbolic functions
with the argument '-x' and to replace it with the
equivalent hyperbolic function with argument x.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -3074,6 +3074,33 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
   }
+  case Intrinsic::cosh: {
+    Value *X;
+    if (match(II->getArgOperand(0), m_OneUse(m_FNeg(m_Value(X))))) {
```
You don't need the one-use check here (See the cos path above).

kshitijvp replied:
Done

kshitijvp commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -3074,6 +3074,33 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
     }
     break;
   }
+  case Intrinsic::cosh: {
```
I don't think so any more specific optimizations are there about hyperbolic functions. So I will merge the paths.

kshitijvp committed:
Merge paths for even and odd functions

Merged the paths for even and odd functions
reducing redundant code and ommited m_OneUse
in the cosh function as it is not required.

kshitijvp commented:
PING!

dtcxzyw commented on llvm/test/Transforms/InstCombine/hyperbolic-functions.ll:
```diff
@@ -0,0 +1,79 @@
+;
+entry:
+  %fneg = fneg double %x
+  %1 = tail call double @llvm.cosh.f64(double %fneg)
```
Use named values (%ret or something else).

kshitijvp replied:
Done

kshitijvp committed:
Added tests for flag preservation

LLVM floating point operations use the Fast-Math Flags
to enable otherwise unsafe floating-point transformations.
Adding tests to check whether these Fast-Math Flags are
preserved across these transformations.

kshitijvp commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -3052,25 +3052,31 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
-      return UnaryOperator::CreateFNegFMF(NewSin, II);
+      // f(-x) --> -f(x)
+      // for f in {sin, sinh, tan, tanh}
+      Value *NewFunc = Builder.CreateUnaryIntrinsic(IID, X, II);
```
Done

dtcxzyw commented on llvm/test/Transforms/InstCombine/hyperbolic-functions.ll:
```diff
@@ -0,0 +1,79 @@
+;
+entry:
+  %fneg = fneg double %x
+  %ret = tail call fast double @llvm.cosh.f64(double %fneg)
```
We don't encourage the usage of fast. Would be better to use more concrete flags (e.g., nnan/ninf/nsz).

dtcxzyw reviewed (approved):
LGTM

kshitijvp committed:
Removed fast flag

Since the use of fast flag is discouraged
I replaced it with the nnan flag.

arsenm commented on llvm/test/Transforms/InstCombine/hyperbolic-functions.ll:
```diff
@@ -0,0 +1,79 @@
+entry:
+  %ret = tail call nnan double @llvm.cosh.f64(double %x)
+  ret double %ret
+}
```
Don't need these no-op tat tests

kshitijvp committed:
Rename tests

Renamed the src tests and removed the redundant
tgt tests
